kernel/src/engine/arrow_expression.rs

```diff
@@ -160,13 +236,24 @@ fn evaluate_expression(
             }
         }
         (Struct(fields), Some(DataType::Struct(schema))) => {
-            let output_schema: ArrowSchema = schema.as_ref().try_into()?;
             let columns = fields
                 .iter()
                 .zip(schema.fields())
                 .map(|(expr, field)| evaluate_expression(expr, batch, Some(field.data_type())));
-            let result =
-                StructArray::try_new(output_schema.fields().clone(), columns.try_collect()?, None)?;
+            let output_cols: Vec<Arc<dyn Array>> = columns.try_collect()?;
+            let output_fields: Vec<ArrowField> = output_cols
+                .iter()
+                .zip(schema.fields())
+                .map(|(array, input_field)| {
+                    let output_data_type =
+                        ensure_data_types(input_field.data_type(), array.data_type());
+                    // use a `map` here so rustc doesn't have to infer the error type
+                    output_data_type.map(|output_type| {
```

Comment:
This is avoiding the double-`?` issue we hit before? 

(but if `ensure_data_types` is really just a validator rather than a mapper... a lot of this code can probably go away?)

Reply:
It's to fix an error like:
```rust
error[E0282]: type annotations needed
   --> kernel/src/engine/arrow_expression.rs:256:21
    |
256 |                     Ok(ArrowField::new(input_field.name(), array.data_type().clone(), array.is_nullable()))
    |                     ^^ cannot infer type of the type parameter `E` declared on the enum `Result`
    |
help: consider specifying the generic arguments
    |
256 |                     Ok::<arrow_schema::Field, E>(ArrowField::new(input_field.name(), array.data_type().clone(), array.is_nullable()))
    |                       ++++++++++++++++++++++++++
```

I've fixed it slightly differently to avoid having to use the turbofish.